File: front/amortizacao-front/src/Financiamento.ts
interface Mes{
  valor: number
  mes: number
}

export class Financiamento {
  private quantidadeDeParcelas = 0;
  private _quantidadeDeParcelasPagas = 0;
  private valorFinanciamento = 0;
  private taxaFinanciamentoEmDecimal = 0;
  private taxaFinanciamentoEmDecimalMensal = 0;
  private _montante = 0;
  private _juros = 0;
  private _jurosPagos = 0;
  private _valorParcelas = new Array<Mes>;
  private _saldoDevedor = 0;
  private parcelaDeAmortizacao = 0;
  constructor(meses: number, valorFinanciamento: number, taxaFinanciamentoEmPorcentagem: number) {
    this.quantidadeDeParcelas = meses;
    this.valorFinanciamento = valorFinanciamento;
    this._saldoDevedor = valorFinanciamento;
    this.taxaFinanciamentoEmDecimal = taxaFinanciamentoEmPorcentagem / 100;
    this.taxaFinanciamentoEmDecimalMensal = this.taxaFinanciamentoEmDecimal / 12;
    this.parcelaDeAmortizacao = valorFinanciamento / this.quantidadeDeParcelas;
    this.calculaJuros();
    this.calculaMontante();
    this.calculaParcelas();
  }

  private calculaParcelas(): void {
    this._valorParcelas = [];
    let quantidadeDeParcelasPagasProv = this._quantidadeDeParcelasPagas;
    while (this.quantidadeDeParcelas > quantidadeDeParcelasPagasProv) {
      let parcela = this.calculaParcela(this.parcelaDeAmortizacao, quantidadeDeParcelasPagasProv,
        this.taxaFinanciamentoEmDecimalMensal,
        this.valorFinanciamento);
      this._valorParcelas.push({valor: parcela, mes: quantidadeDeParcelasPagasProv});
      quantidadeDeParcelasPagasProv += 1;
    }
  }

  public calculaParcela(valorAmortizacao: number, quantidadeDeParcelasPagas: number,
    taxaDeJurosMensal: number, valorFinanciamento: number): number {
    // parcela = valor_da_amortizacao + (taxa_de_juros_mensal*(valor_financiamento -
    // (quantidade_de_parcelas_pagas * valor_amortização)))
    let parcela = valorAmortizacao + (taxaDeJurosMensal * (
      valorFinanciamento - (quantidadeDeParcelasPagas * valorAmortizacao)));
    return parcela;
  }


  private calculaMontante(): void {
    this._montante = this._saldoDevedor + this._juros;
  }

  private calculaJuros(): void {
    this._juros = this._saldoDevedor * this.taxaFinanciamentoEmDecimalMensal * 100;
  }
  private amortizaUmaParcela() {
    this._saldoDevedor = this._saldoDevedor - this.parcelaDeAmortizacao;
    this._quantidadeDeParcelasPagas++;
  }
  public amortiza(numeroDeParcelas: number) {
    if (!(numeroDeParcelas > (this.quantidadeDeParcelas - this._quantidadeDeParcelasPagas))) {
      for (let i = 0; i < numeroDeParcelas; i++) {
        this.amortizaUmaParcela();
      }
      this.recalcula();
    }
  }


  public pagaUmaParcela() {
    if (this._quantidadeDeParcelasPagas < this.quantidadeDeParcelas) {
      this.amortizaUmaParcela();
      this._jurosPagos = this._jurosPagos + (this._valorParcelas[0].valor - (this.parcelaDeAmortizacao));
      this.recalcula();
    }
  }

  private recalcula() {
    this.calculaJuros();
    this.calculaMontante();
    this.calculaParcelas();
  }

  public get montante(): number {
    return this._montante;
  }

  public get juros(): number {
    return this._juros
  }
  public get jurosPagos(): number {
    return this._jurosPagos
  }

  public get valorParcelas(): Array<Mes> {
    return this._valorParcelas
  }

  public get saldoDevedor(): number {
    return this._saldoDevedor
  }

  public get quantidadeDeParcelasPagas(): number {
    return this._quantidadeDeParcelasPagas
  }




}

File: java/src/main/java/br/com/projeto/amortizacao/Financiamento.java
package br.com.projeto.amortizacao;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;

public class Financiamento {
  private int quantidadeDeParcelas;
  private int quantidadeDeParcelasPagas;
  private BigDecimal valorFinanciamento = BigDecimal.valueOf(0);
  private BigDecimal taxaFinanciamentoEmDecimal = BigDecimal.valueOf(0);
  private BigDecimal taxaFinanciamentoEmDecimalMensal = BigDecimal.valueOf(0);
  private BigDecimal montante = BigDecimal.valueOf(0);
  private BigDecimal juros = BigDecimal.valueOf(0);
  private BigDecimal jurosPagos = BigDecimal.valueOf(0);
  private List<BigDecimal> valorParcelas = new ArrayList<>();
  private BigDecimal saldoDevedor = BigDecimal.valueOf(0L);

  private BigDecimal parcelaDeAmortizacao = BigDecimal.valueOf(0L);

  public Financiamento(int meses, BigDecimal valorFinanciamento, BigDecimal taxaFinanciamentoEmPorcentagem) {
    this.quantidadeDeParcelas = meses;
    this.valorFinanciamento = valorFinanciamento;
    this.saldoDevedor = valorFinanciamento;
    this.taxaFinanciamentoEmDecimal = taxaFinanciamentoEmPorcentagem.divide(BigDecimal.valueOf(100L),
        RoundingMode.HALF_UP).setScale(4);
    this.taxaFinanciamentoEmDecimalMensal = taxaFinanciamentoEmDecimal
        .divide(BigDecimal.valueOf(12L), RoundingMode.HALF_UP).setScale(4);
    this.parcelaDeAmortizacao = valorFinanciamento
        .divide(BigDecimal.valueOf(Integer.toUnsignedLong(quantidadeDeParcelas)), RoundingMode.HALF_UP);
    calculaJuros();
    calculaMontante();
    calculaParcelas();
  }

  private void calculaParcelas() {
    this.valorParcelas.removeAll(this.valorParcelas);
    long quantidadeDeParcelasPagasProv = quantidadeDeParcelasPagas;
    while (this.quantidadeDeParcelas > quantidadeDeParcelasPagasProv) {
      BigDecimal parcela = calculaParcela(this.parcelaDeAmortizacao, quantidadeDeParcelasPagasProv,
          this.taxaFinanciamentoEmDecimalMensal,
          this.valorFinanciamento);
      this.valorParcelas.add(parcela.setScale(2, RoundingMode.HALF_UP));
      quantidadeDeParcelasPagasProv += 1;
    }
  }

  public BigDecimal calculaParcela(BigDecimal valorAmortizacao, long quantidadeDeParcelasPagas,
      BigDecimal taxaDeJurosMensal, BigDecimal valorFinanciamento) {
    // parcela = valor_da_amortizacao + (taxa_de_juros_mensal*(valor_financiamento -
    // (quantidade_de_parcelas_pagas * valor_amortização)))
    BigDecimal parcela = valorAmortizacao.add(taxaDeJurosMensal.multiply(
        valorFinanciamento.subtract(BigDecimal.valueOf(quantidadeDeParcelasPagas).multiply(valorAmortizacao))));
    return parcela;
  }

  public void amortiza(long numeroDeParcelas) {
    if (!(numeroDeParcelas > (quantidadeDeParcelas - quantidadeDeParcelasPagas))) {
      for (int i = 0; i < numeroDeParcelas; i++) {
        amortiza();
      }
      recalcula();
    }
  }

  private void amortiza() {
    this.saldoDevedor = this.saldoDevedor.subtract(this.parcelaDeAmortizacao);
    quantidadeDeParcelasPagas++;
  }

  public void pagaUmaParcela() {
    if (quantidadeDeParcelasPagas < quantidadeDeParcelas) {
      amortiza();
      jurosPagos = jurosPagos.add(valorParcelas.get(0).subtract(parcelaDeAmortizacao));
      recalcula();
    }
  }

  private void recalcula() {
    calculaJuros();
    calculaMontante();
    calculaParcelas();
  }

  private void calculaMontante() {
    this.montante = this.saldoDevedor.add(this.juros);
  }

  private void calculaJuros() {
    this.juros = this.saldoDevedor.multiply(this.taxaFinanciamentoEmDecimalMensal.multiply(BigDecimal.valueOf(100L)));
  }

  public BigDecimal getMontante() {
    return montante.setScale(2, RoundingMode.HALF_UP);
  }

  public BigDecimal getJuros() {
    return juros.setScale(2, RoundingMode.HALF_UP);
  }

  public List<BigDecimal> getValorParcelas() {
    System.out.println("Parcelas atuais: " + this.valorParcelas.size());
    return valorParcelas;
  }

  public BigDecimal getSaldoDevedor() {
    return saldoDevedor;
  }

  public int getQuantidadeDeParcelasPagas() {
    return quantidadeDeParcelasPagas;
  }

  public BigDecimal getJurosPagos() {
    return jurosPagos;
  }
}
